src/Common/Internal/Authentication/SharedAccessSignatureHelper.php

```diff
@@ -0,0 +1,277 @@
+
+        // construct an array with the parameters to generate the shared access signature at the account level
+        $parameters = array();
+        $parameters[] = $this->accountName;
+        $parameters[] = urldecode($signedPermissions);
+        $parameters[] = urldecode($signedService);
+        $parameters[] = urldecode($signedResourceType);
+        $parameters[] = urldecode($signedStart);
+        $parameters[] = urldecode($signedExpiracy);
+        $parameters[] = urldecode($signedIP);
+        $parameters[] = urldecode($signedProtocol);
+        $parameters[] = urldecode($signedVersion);
+
+        // implode the parameters into a string
+        $stringToSign = utf8_encode(implode("\n", $parameters)."\n");
+
+        // decode the account key from base64
+        $decodedAccountKey = base64_decode($this->accountKey);
+        
+        // create the signature with hmac sha256
+        $signature = hash_hmac("sha256", $stringToSign, $decodedAccountKey, true);
+
+        // encode the signature as base64
+        $base64Signature = base64_encode($signature);
+
+        // return the signature
+        return $base64Signature;
+    }
+
+    function validateAndSanitizeSignedService($signedService) {
+        // validate signed service is not null or empty
+        Validate::isString($signedService, 'signedService');
+        Validate::notNullOrEmpty($signedService, 'signedService');
+
+        // sanitize signed service
+        $sanitizedSignedService = $this->removeDuplicateCharacters(strtolower($signedService));
+        
+        // The signed service should only be a combination of the letters b(lob) q(ueue) t(able) or f(ile)
+        $signedServiceIsValid = preg_match("/^[bqtf]*$/", $sanitizedSignedService);
+        if(!$signedServiceIsValid) {
+            throw new \InvalidArgumentException(Resources::SIGNED_SERVICE_INVALID_VALIDATION_MSG);
+        }
+
+        return $sanitizedSignedService;
+    }
+
+    function validateAndSanitizeSignedResourceType($signedResourceType) {
+        // validate signed resource type is not null or empty
+        Validate::isString($signedResourceType, 'signedResourceType');
+        Validate::notNullOrEmpty($signedResourceType, 'signedResourceType');
+
+        // sanitize signed resource type
+        $sanitizedSignedResourceType = $this->removeDuplicateCharacters(strtolower($signedResourceType));
+        
+        // The signed resource type should only be a combination of the letters s(ervice) c(container) or o(bject)
+        $signedResourceTypeIsValid = preg_match("/^[sco]*$/", $sanitizedSignedResourceType);
+        if(!$signedResourceTypeIsValid) {
+            throw new \InvalidArgumentException(Resources::SIGNED_RESOURCE_TYPE_INVALID_VALIDATION_MSG);
+        }
+
+        return $sanitizedSignedResourceType;
+    }
+
+    function validateAndSanitizeSignedPermissions($signedPermissions, $signedService, $signedResourceType) {
```

Comment:
Should still validate if signed permission consist only of characters 'rwdlacup'. E.g. 'rwdlacupzx' would fail the validation but 'rrr' would pass. Also reordering is necessary.